Prevent leaking seclist ingress and egress rules when updating a service

templecloud opened the pull request:
This commit fixes a particular type of seclist rule leak in the CCM. When a services port or health check port is mutated then the old/original _'single-range' port rules_ created in the 'CCM ingress' and 'K8sWorker egress' seclists (or equivalent) are not deleted. This can be replicated consistently by running the 'Canary' e2e tests that performs a mutation test on a service. This fix also includes small optimisation to only create seclist rules for a loadbalancer after it has been successfully created.

This commit includes the following:

1. Two small conservative fixes to clean up the leaking ingress/egress rules when a service port is mutated. This required a few small alterations to pass up existing port information to the 'getEgressRules()' functions.

2. Some additional unit tests to check the above fixes.

3. Added an OCI client to the e2e test framework and added a simple sanity check to count the number seclist rules for a port before and after mutation. To keep things simple this test only runs if configured to do so. Configuring the tests consists of specifying the OCID of the target seclists to check as environment variables.

This commit does not include any refactoring or tidying up of these existing seclist rule management functionality. I think this could be harmonised a little and made more clear; then surrounded with further tests. In particular the `getLoadBalancerEgressRules()` and `getNodeIngressRules() `are related but work differently. I rolled back such changes for this commit as there was other ongoing work on seclists.

Anyway, please check carefully; this is my first foray into this stuff and needed a fun rebase to get the new changes to logging etc.

templecloud commented on Makefile:
```diff
@@ -29,7 +29,7 @@ SRC_DIRS := cmd pkg # directories which hold app source (not vendored)
 # Allows overriding where the CCM should look for the cloud provider config
 # when running via make run-dev.
-CLOUD_PROVIDER_CFG ?= $$(pwd)/cloud-provider.yaml
+CLOUDCONFIG ?= $$(pwd)/cloud-provider.yaml
```
Need to remember why I wanted to change this. Can change back.

templecloud commented on hack/test-e2e.sh:
```diff
@@ -26,10 +26,7 @@ function run_e2e_tests() {
-        --kubeconfig="${KUBECONFIG}" \
-        --cloud-config="${CLOUDCONFIG}" \
-        --delete-namespace=false
+        -- --kubeconfig=${KUBECONFIG} --cloud-config=${CLOUDCONFIG} --delete-namespace=true
```
Agh, the change in the Makefile above might have been to match this usage in the tests. I can still roll it back.

owainlewis replied:
It seems irrelevant to this PR so would prefer it to be reversed to simplify review.

templecloud replied:
I rolled this back, but reckon we should have kept it. No-one will fix it in a single PR and in any other PR the same argument can be made! Anyway, should be reverted now!

templecloud commented on pkg/oci/ccm.go:
```diff
@@ -235,9 +235,9 @@ func buildConfigurationProvider(logger *zap.Logger, config *Config) (common.Conf
 }
 
-// newRateLimiter builds and returns a struct containing read and write
+// NewRateLimiter builds and returns a struct containing read and write
```
I needed this when creating the new OCI client for the e2e tests. It seemed wrong to cut and paste the code. This function seems to be a 'constructor', so I made it public.

templecloud commented on pkg/oci/load_balancer.go:
```diff
@@ -293,7 +287,18 @@ func (cp *CloudProvider) createLoadBalancer(ctx context.Context, spec *LBSpec) (
 
 	logger.With("loadBalancerID", *lb.Id).Info("Load balancer created")
-	return loadBalancerToStatus(lb)
+	status, err := loadBalancerToStatus(lb)
```
I moved adding seclists to after we have created the loadbalancer. If creation fails for some reason the rules will not be added. Could there be unforseen issues with this?

templecloud commented on pkg/oci/load_balancer_security_lists.go:
```diff
@@ -152,12 +152,20 @@ func (s *baseSecurityListManager) updateLoadBalancerRules(ctx context.Context, l
 
-		lbEgressRules := getLoadBalancerEgressRules(logger, secList.EgressSecurityRules, nodeSubnets, ports.BackendPort, s.serviceLister)
-		lbEgressRules = getLoadBalancerEgressRules(logger, lbEgressRules, nodeSubnets, ports.HealthCheckerPort, s.serviceLister)
+		// -1 denotes nil ports.
```
I need to thread though the 'current' backEnd and health ports in-order to carefully delete the leaking rules. As I mentioned at some point I would like to tidy up this double invocation to `getLoadBalancerEgressRules()` and make it work more like ` getNodeIngressRules()`.

templecloud commented on pkg/oci/load_balancer_security_lists.go:
```diff
@@ -362,6 +378,23 @@ func getNodeIngressRules(
 	for _, rule := range rules {
+		// Remove (do not re-add) any rule that represents the old case when
+		// mutating a single ranged backend port or health check port.
+		if rule.TcpOptions != nil && rule.TcpOptions.DestinationPortRange != nil &&
```
This part of the fix. I wanted to be on the safe side so made it very restrictive. It should only target the 'single port rule mutation' described.

templecloud commented on pkg/oci/load_balancer_security_lists.go:
```diff
@@ -531,8 +564,21 @@ func getLoadBalancerEgressRules(
 
 	egressRules := []core.EgressSecurityRule{}
 	for _, rule := range rules {
+		// Remove (do not re-add) any rule that represents the old case when mutating a single ranged port.
```
This is the second part of the fix. I wanted to be on the safe side so made it very restrictive. It should only target the 'single port rule mutation' described.

templecloud commented on pkg/oci/load_balancer_security_lists_test.go:
```diff
@@ -195,6 +195,37 @@ func TestGetNodeIngressRules(t *testing.T) {
 				makeIngressSecurityRule("1", 80),
 				makeIngressSecurityRule("2", 80),
 			},
+		}, {
```
Unit test for health check port mutation. I chose to put more realistic strings in. IMHO it makes these tests easier to read/understand, especially if we added further input rules in. I can make it like the other though.

templecloud commented on pkg/oci/load_balancer_security_lists_test.go:
```diff
@@ -382,13 +415,90 @@ func TestGetLoadBalancerEgressRules(t *testing.T) {
 				makeEgressSecurityRule("1", 80),
 				makeEgressSecurityRule("2", 80),
 			},
+		}, {
```
Unit test for service port mutation. I chose to put more realistic strings in. IMHO it makes these tests easier to read/understand, especially if we added further input rules in. I can make it like the other though.

templecloud commented on test/e2e/framework/framework.go:
```diff
@@ -64,13 +69,20 @@ type Framework struct {
 	ClientSet         clientset.Interface
 	InternalClientset internalclientset.Interface
 
+	CloudProviderConfig *oci.Config      // If specified, the CloudProviderConfig. This provides information on the configuration of the test cluster.
```
Configurable OCI client to run some seclist sanity checks.

templecloud commented on test/e2e/framework/framework.go:
```diff
@@ -226,3 +251,30 @@ func (f *Framework) AfterEach() {
+}
+
+// createOCIClient creates an OCI client derived from the CCM's cloud provider config file.
+func createOCIClient(cloudProviderConfig *oci.Config) (client.Interface, error) {
```
I did not see the point in adding configuration files. We grab the auth info from the cloud config.

templecloud commented on test/e2e/framework/seclist_utils.go:
```diff
@@ -0,0 +1,143 @@
+/*
```
A set of utility functions that count the number of ingress/egress port rules in the specified seclists. Used by the test check the right number of rules are mutated and all the old rules are deleted.

templecloud commented on pkg/oci/load_balancer_security_lists.go:
```diff
@@ -238,7 +246,7 @@ func (s *defaultSecurityListManager) Delete(ctx context.Context, lbSubnets []*co
 	noSourceCIDRs := []string{}
 
-	err := s.updateLoadBalancerRules(ctx, lbSubnets, noSubnets, noSourceCIDRs, ports)
+	err := s.updateLoadBalancerRules(ctx, lbSubnets, noSubnets, noSourceCIDRs, &ports, ports)
```
Chose sending in port twice over nil. I think this makes sense.

owainlewis commented on hack/boilerplate/boilerplate.py:
```diff
@@ -146,6 +146,7 @@ def file_extension(filename):
     "test/e2e/load_balancer.go",
     "test/e2e/framework/cleanup.go",
     "test/e2e/framework/framework.go",
+    "test/e2e/framework/seclist_utils.go",
```
seclist_utils => seclist_util to be consistent with test/e2e/framework/service_util.go

templecloud replied:
It was matching actually file called 'network_utils'. I made all util files singular.

owainlewis commented on pkg/oci/load_balancer_security_lists.go:
```diff
@@ -143,7 +143,7 @@ func (s *baseSecurityListManager) updateBackendRules(ctx context.Context, lbSubn
 // updateLoadBalancerRules handles updating the ingress and egress rules for the load balance subnets.
 // If the listener is nil, then only egress rules from the load balancer to the backend subnets will be checked.
-func (s *baseSecurityListManager) updateLoadBalancerRules(ctx context.Context, lbSubnets []*core.Subnet, nodeSubnets []*core.Subnet, sourceCIDRs []string, ports portSpec) error {
+func (s *baseSecurityListManager) updateLoadBalancerRules(ctx context.Context, lbSubnets []*core.Subnet, nodeSubnets []*core.Subnet, sourceCIDRs []string, actualPorts *portSpec, desiredPorts portSpec) error {
```
We have actualPorts and desiredPorts here but one is a pointer, the other isn't. Is there any reason?

templecloud replied:
Yeagh, I agree - darn fuggly init? 

To cut a long story short I think it derives from another bit of fuggly + typo on master (here)[https://github.com/oracle/oci-cloud-controller-manager/blob/e11429bc410ec62e52ba0aa504b777677bd3f908/pkg/oci/load_balancer_security_lists.go#L75].

I didn't fix it up on this run because I was trying to keep the changes small and close to the problem I am attempting to fix - as requested. It is really just cosmetic. I am happy to fix it up in this commit or as a subsequent one. Up to you. Please advise.

owainlewis replied:
I was just curious in case it was a typo. If there are call sites that require nil values then we can leave it as is.

owainlewis commented on test/e2e/framework/framework.go:
```diff
@@ -95,7 +107,10 @@ func NewFramework(baseName string, client clientset.Interface) *Framework {
 	}
-
+	// Dev/CI only configuration. The seclist for CCM load-balancer routes.
+	f.CCMSecListID = os.Getenv("CCM_SECLIST_ID")
```
Both these values will default to an empty string if the env var is not set. Would that cause logical problems when passed to framework.CountSecListTunnelSinglePortRules ?

templecloud replied:
A good thing to query, but, hopefully it is ok. I hope it is the case that if these seclists values are not set then the [test calculations](https://github.com/oracle/oci-cloud-controller-manager/blob/1e79b88e7f330a0348e6aa46cd84cf621259c9cf/test/e2e/framework/seclist_util.go#L65) will collapse into 0 - so they will reported a s valid. I tested it manually and it seemed to work ok. We could make it explicit, but, it would make the tests more verbose to read.

templecloud replied:
Done.

owainlewis commented on pkg/oci/load_balancer_security_lists.go:
```diff
@@ -362,6 +378,23 @@ func getNodeIngressRules(
+		// mutating a single ranged backend port or health check port.
+		if rule.TcpOptions != nil && rule.TcpOptions.DestinationPortRange != nil &&
+			*rule.TcpOptions.DestinationPortRange.Min == *rule.TcpOptions.DestinationPortRange.Max &&
+			*rule.TcpOptions.DestinationPortRange.Min != desiredPorts.BackendPort && *rule.TcpOptions.DestinationPortRange.Max != desiredPorts.BackendPort &&
```
This large logical conjunction looks quite prone to mistakes and I think is duplicated again below. Might be worth factoring this into a smaller helper function and unit testing.

templecloud replied:
Yeagh, I mentioned this in the JIRA and I started something like this on the ('other branch')[https://github.com/oracle/oci-cloud-controller-manager/blob/f2a9f78c50cb231128c27ee23392db84bca0749e/pkg/oci/load_balancer_security_lists.go#L516]. However, I thought we were meant to be making this fix with a few changes possible and then iterating to improve other bits? I am happy to have a go at improving this now, but, was trying to avoid a 'large commit'. Do you want me to refactor this code in this branch along with the fix attempt?

templecloud commented:
Thanks for the review OL! You have suggested several good changes that I was originally going to raise a JIRA for and do as a separate bit work. Can you please confirm you want me to do it here and now? Then I will get cracking!

templecloud commented on test/e2e/framework/seclist_util.go:
```diff
@@ -0,0 +1,143 @@
+
+// CountSecListTunnelSinglePortRules counts the number of 'single port'
+// (non-ranged) egress/ingress rules for the specified list and port.
+func CountSecListTunnelSinglePortRules(oci client.Interface, egressSecListID, ingressSecListID string, port int) (int, int) {
```
Thought this might come up! Just wanted to mention I preferred environment variables here as I only intended these tests to be executed by our CCM ci pipeline (where we manage the seclist ids). Not by the canary tests etc where we don't. I felt threading them through was a bit noisy. Still if you want flags for consistency ill do it.

templecloud commented:
Thanks for the review guys. Made all changes. Also added some additional e2e configuration to the wercker yaml after chatting with @prydie. Squashed all commits.

kristenjacobs commented:
The logic all looks fine to me.

prydie commented:
🎉
